File: discord_bot/slash_commands/check_repos.py
import logging

import discord
from discord import app_commands

from discord_bot.discord_bot import SggBot
from goon_github.repositories import get_all_repos

log = logging.getLogger(__name__)

SYSOP_ROLE_ID = 907314874101665823


def check_if_user_in_role(role_to_check_for: int, roles: list[discord.Role]) -> bool:
    for role in roles:
        if role.id == role_to_check_for:
            return True
    return False


def add_check_repos_slash_command(bot: SggBot, guild: discord.Object):
    @bot.tree.command(name="check-repos", description="Gets a list of 10 of the repos by kjb, starting at the first", guild=guild)
    @app_commands.describe(start_number="The number to start listing repositories from")
    async def check_repos(interaction: discord.Interaction, start_number: int = 0):
        assert isinstance(interaction.user, discord.Member)
        roles = interaction.user.roles
        if not check_if_user_in_role(SYSOP_ROLE_ID, roles):
            await interaction.response.send_message("Only Sysop can perform this action")
            return

        await interaction.response.defer()
        repos = get_all_repos()
        end = start_number + 10

        embed = discord.Embed(title="Repos")
        for i in range(start_number, min(end, len(repos))):
            repo = repos[i]
            embed.add_field(name=repo.name, value=repo.default_branch or "main", inline=False)

        await interaction.followup.send(content="Heres a list of 10 the Repos!", embed=embed)

File: discord_bot/slash_commands/check_repos.go
package slash_commands

import (
	"log"

	"github.com/bwmarrin/discordgo"

	"sgg-discord-bot/goon_github"
)

const SysopRoleID = "907314874101665823"

var CheckReposCommand = &discordgo.ApplicationCommand{
	Name:        "check-repos",
	Description: "Gets a list of 10 of the repos by kjb, starting at the first",
	Options: []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionInteger,
			Name:        "start_number",
			Description: "The number to start listing repositories from",
			Required:    false,
		},
	},
}

func checkIfUserInRole(roleToCheck string, roles []string) bool {
	for _, role := range roles {
		if role == roleToCheck {
			return true
		}
	}
	return false
}

func HandleCheckRepos(s *discordgo.Session, i *discordgo.InteractionCreate) {
	if !checkIfUserInRole(SysopRoleID, i.Member.Roles) {
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: "Only Sysop can perform this action",
			},
		})
		return
	}

	s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseDeferredChannelMessageWithSource,
	})

	startNumber := int64(0)
	options := i.ApplicationCommandData().Options
	for _, opt := range options {
		if opt.Name == "start_number" {
			startNumber = opt.IntValue()
		}
	}

	repos := goon_github.GetAllRepos()
	end := int(startNumber) + 10

	var fields []*discordgo.MessageEmbedField
	for idx := int(startNumber); idx < end && idx < len(repos); idx++ {
		repo := repos[idx]
		branch := "main"
		if repo.GetDefaultBranch() != "" {
			branch = repo.GetDefaultBranch()
		}
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   repo.GetName(),
			Value:  branch,
			Inline: false,
		})
	}

	embed := &discordgo.MessageEmbed{
		Title:  "Repos",
		Fields: fields,
	}

	_, err := s.FollowupMessageCreate(i.Interaction, true, &discordgo.WebhookParams{
		Content: "Heres a list of 10 the Repos!",
		Embeds:  []*discordgo.MessageEmbed{embed},
	})
	if err != nil {
		log.Printf("Error sending followup: %v", err)
	}
}
